parlai/utils/torch.py

```diff
@@ -341,8 +341,15 @@ def make_parallel(self, model: torch.nn.Module) -> torch.nn.Module:
         self.__device_allocations['cuda:0'] += trainable_parameters(model) * 3
 
         model.apply(self._place_modulelist)
+        model._apply(self._move_rest_to_cuda0)
         return model
 
+    def _move_rest_to_cuda0(self, parameter):
+        if parameter.device.type == 'cpu':
+            return parameter.to('cuda:0')
```

Comment:
Hm can you walk me through this? This looks like it is moving at least some weights just to GPU 0

Reply:
Yes, it's moving everything that wasn't explicitly allocated to another GPU already to GPU0. The way I implemented model parallel was that I shard the layers (unevenly) across all the GPUs, but left things like the embedding weights on GPU0 (which is why I was uneven). Since the goal was to avoid initializing everything on one GPU, I move everything special around first, and then search for anything leftover.